libethereum/EthereumHost.cpp

```diff
@@ -218,25 +218,32 @@ void EthereumHost::maintainTransactions()
 		}
 }
 
-pair<vector<shared_ptr<EthereumPeer>>, vector<shared_ptr<EthereumPeer>>> EthereumHost::randomSelection(unsigned _percent, std::function<bool(EthereumPeer*)> const& _allow)
+tuple<vector<shared_ptr<EthereumPeer>>, vector<shared_ptr<EthereumPeer>>, list<shared_ptr<Session>>> EthereumHost::randomSelection(unsigned _percent, std::function<bool(EthereumPeer*)> const& _allow)
 {
-	pair<vector<shared_ptr<EthereumPeer>>, vector<shared_ptr<EthereumPeer>>> ret;
-	ret.second.reserve(peerSessions().size());
-	for (auto const& j: peerSessions())
+	vector<shared_ptr<EthereumPeer>> chosen;
+	vector<shared_ptr<EthereumPeer>> allowed;
+	list<shared_ptr<Session>> sessions;
+	
+	auto const& ps = peerSessions();
+	allowed.reserve(ps.size());
+	for (auto const& j: ps)
 	{
 		auto pp = j.first->cap<EthereumPeer>();
 		if (_allow(pp.get()))
-			ret.second.push_back(pp);
+		{
+			allowed.push_back(move(pp));
+			sessions.push_back(move(j.first));
+		}
 	}
 
-	ret.second.reserve((peerSessions().size() * _percent + 99) / 100);
-	for (unsigned i = (peerSessions().size() * _percent + 99) / 100; i-- && ret.second.size();)
+	chosen.reserve((ps.size() * _percent + 99) / 100);
+	for (unsigned i = (ps.size() * _percent + 99) / 100; i-- && allowed.size();)
 	{
-		unsigned n = rand() % ret.second.size();
-		ret.first.push_back(std::move(ret.second[n]));
-		ret.second.erase(ret.second.begin() + n);
+		unsigned n = rand() % allowed.size();
+		chosen.push_back(std::move(allowed[n]));
+		allowed.erase(allowed.begin() + n);
 	}
-	return ret;
+	return make_tuple(chosen, allowed, sessions);
```

Comment:
That's somehow correct. make_tuple moves arguments but only if you pass a rvalue reference. Currently you're passing lvalue references as all arguments have a name. To make them a rvalue reference you need to cast. The cast instruction is just `std::move`.